src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs

```diff
@@ -1460,24 +1461,41 @@ public async Task WaitForExitAsync(CancellationToken cancellationToken = default
 
             var tcs = new TaskCompletionSourceWithCancellation<bool>();
 
-            EventHandler handler = (s, e) => tcs.TrySetResult(true);
+            EventHandler handler = (_, _) => tcs.TrySetResult(true);
             Exited += handler;
 
             try
             {
                 if (HasExited)
                 {
                     // CASE 1.2 & CASE 3.2: Handle race where the process exits before registering the handler
-                    return;
+                }
+                else
+                {
+                    // CASE 1.1 & CASE 3.1: Process exits or is canceled here
+                    await tcs.WaitWithCancellationAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
It's strange we're now using two different "with cancellation" waiting approaches in the same body of code.

Reply:
I noticed that as well. While there are similarities I don't think we should combine the two approaches.